Feature/tinymce bridge js settings

spatarel opened the pull request:
Here are some code changes I made so that the issues presented in #668 are fixed.

The trick I used was to create 2 new JS files, one for Chosen, the other for Tiny MCE, storing an empty JS map each.
These 2 JS maps are then used by the JS code to instantiate Chosen with the specified options and to expand the Tiny MCE settings loaded from the `tinymce.yml` file.

At first glance, this has no effect, since the 2 JS maps are empty. However, since they each reside in an empty file, these files can be overloaded by the user and he/she can freely populate the JS hashes with the custom settings required by their project.

Please take the time to test the code.
Thank you!

scambra reviewed (changes requested):
Can you use ActiveScaffold.js_config instead of chosen_options and tiny_mce_settings? You may remove the new files, and use ActiveScaffold.js_config.chosen_options || {} and ActiveScaffold.js_config.tiny_mce_settings || {} instead of those global variables

ActiveScaffold.js_config can be defined in an initializer

spatarel commented:
Unfortunately, I wasn't able to move the configuration code in ruby, but not for a lack of trying.

Here is what I tried:
```
my_project/app/controllers/application_controller.rb:

  before_action :configure_active_scaffold

  def configure_active_scaffold
    ActiveScaffold.js_config[:chosen_options] = {
      width: '1000px',
      max_shown_results: '100',
      search_contains: true
    }
    ActiveScaffold.js_config[:tiny_mce_settings] = {
      key: "value",
      key2: "value2",
      key3: "value3",
      key4: "value4"
    }
  end
```

However, I encountered 3 main issues that I couldn't overcome:

(1) The JS callback function needs to be written in a `.js` (or `.js.erb`) file. It must be written as JavaScript code.
The only way to avoid this would be to use some JS native function that converts a string to a function. But this would require a lot of work. And there's the issue of how to tell if a string parameter was intended as a string or as a function.

(2) Whenever I change the settings in `configure_active_scaffold`, they are not taken into account because the JavaScript code is not recompiled.
This means that whenever one needs to change the settings/configurations they must also somehow force the JS code to recompile. I was able to force this by updating a `.js` file in `/app/assets/javascripts/`. But... this defies the purpose...

(3) I was left with setting up `ActiveScaffold.config.chosen_options` and `ActiveScaffold.config.tiny_mce_settings` from some `.js` file in `/app/assets/javascripts/` However, not even this seemed to be working well, because `ActiveScaffold.config` was `undefined`. Therefore, I had to use `ActiveScaffold.chosen_options` and `ActiveScaffold.tiny_mce_settings` respectively instead.

What should I do?

I would suggest the inclusion of some empty `.js` file such as: `active_scaffold_config.js`.
Then the user can override it:
```
/app/assets/javascripts/active_scaffold_config.js:

ActiveScaffold.chosen_options = {
  width: '1000px',
  max_shown_results: 100,
  search_contains: true
};

ActiveScaffold.tiny_mce_settings = {
  init_instance_callback: function (editor) {
    editor.shortcuts.add('f9', 'Preview', function () {
      editor.execCommand('mcePreview');
    });
  }
};
```

scambra commented:
1. You can't use before_action to set js_config, it must be set in an initializer, so it's applied when assets are compiled.

2. If you want to use different js settings in every request you must add some JavaScript in your view or layout, to change ActiveScaffold.config.chosen_options. I don't remember if tinymce bridge support to set some settings in column.options, I will review it, but I think it supports it, I can see js code reads data('tinymce') probably form helper sets data-tinymce attribute.

3. I think you should use ActiveScaffold.config.chosen_options, but I will test your PR in next few days, as soon as I can.
3.

spatarel commented:
I think I made it work!

As you sugested in (2), I moved the JS configuration code in my `<head>` tag, from my layout file. Then everything started working without any asset precompilation needed. `ActiveScaffold.config` was defined. Any changes to the configurations were applied after a simple page refresh. Of course, there was no longer the need to include any JS config files (my silly idea :P).

Thus, in my latest commits:
(1) I removed the inclusion of the JS config files.
(2) I deleted the JS config files.
(3) I reverted to using `ActiveScaffold.config.chosen_options` and `ActiveScaffold.config.tiny_mce_settings`.

Here is my `<head>` tag as an example. Feel free to add my JS configuration code as an example to the wiki, for future reference.
```
<head>
  <!-- Configure ActiveScaffold JS -->
  <script type="text/javascript">
    ActiveScaffold.config.chosen_options = {
      width: '1000px',
      max_shown_results: 100,
      search_contains: true
    };

    ActiveScaffold.config.tiny_mce_settings = {
      init_instance_callback: function (editor) {
        editor.shortcuts.add('f9', 'Preview', function () {
          editor.execCommand('mcePreview');
        });
      }
    };
  </script>
</head>
```

Feel free to test my work and let me know if I got it right or it needs more fixing.
Thank you for this gem! (`active_scaffold` really is a gem!)

scambra commented:
The JS in head is the way to setup a tinymce callback, as it isn't possible to set in ruby code which is converted to json. It should work too if you create a js file with that code and add it to your manifest (application.js) after require active_scaffold.

Also, settings for chosen should work with this initializer (e.g. config/initializers/active_scaffold.rb):

```ruby
ActiveScaffold.js_config[:chosen_options] = {
      width: '1000px',
      max_shown_results: '100',
      search_contains: true
    }
```

Can you test it?
